### PR TITLE
Create account

### PR DESCRIPTION
**Pull request by wSedlacek**
# Description
Implements the required routes and middleware to handle login requires on the backend.

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## Change Status

- [x] Complete, tested, ready to review and merge
- [x] Complete, but not tested (may need new tests)
- [x] Incomplete/work-in-progress, PR is for discussion/feedback

# How Has This Been Tested?

- [x] Backend Testing for Validation
- [x] Backend Testing for Route response

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My code has been reviewed by at least one peer
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] There are no merge conflicts

### CODE REVIEW

**Comment by wSedlacek** [comment]
[[Parent] - Create an account](https://trello.com/c/3zKsookg/19-parent-create-an-account)

**Comment by wSedlacek** [comment]
If this could be reviewed that would be awesome.
Let me know if there are any areas that you would like to see comments.

**Review by ChrisJCarter91 — APPROVED** [review]
All code looks good, follows style guide, and commit messages make it easy to follow. Good work.

File: src/index.ts [hunk 1]
```diff
@@ -1,12 +1,21 @@
...
+createConnection(connection()).then(async () => {
     const app = express();
-    app.use(exampleRoutes);
+    globalMiddleware(app);
+
+    app.use('/auth', authRoutes);
 
-    app.listen(4000);
-    console.log(`Listening on port ${4000}`);
+    const port = process.env.PORT;
```
Comment (EricBanker12) [inline src/index.ts]: Locally, port is undefined. Should this be `const port = process.env.PORT | 4000;`?

File: src/middleware/hash/hash.middleware.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import { getRepository } from 'typeorm';
+
+import { Middleware } from '../../models/common/Middleware';
+import { Parent } from '../../database/entity/Parent';
+import { connection } from '../../util/typeorm-connection';
+
+const Hash: Middleware = () => async (req, res, next) => {
+    try {
+        if (!req.register) throw new Error('No register object on body...');
+        const salt: number = parseInt(process.env.SALT, 10);
```
Comment (EricBanker12) [inline src/middleware/hash/hash.middleware.ts]: I believe `process.env.SALT` is undefined. should it have a default for development?

File: src/routes/auth/auth.routes.spec.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    }),
+});
+
+describe('POST /login', () => {
+    it('should return 200 when logging in correctly', async () => {
+        await request(app)
+            .post('/auth/login')
+            .send({ username: 'Test@mail.com', password: 'Test1234' })
+            .expect(200);
+    });
```
Comment (EricBanker12) [inline src/routes/auth/auth.routes.spec.ts]: expected 200 "OK", got 500 "Internal Server Error"

File: src/routes/auth/auth.routes.spec.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+});
+
+describe('POST /register', () => {
+    it('should return 201 when registration is completed', async () => {
+        await request(app)
+            .post('/auth/register')
+            .send({ username: 'NewTest@gmail.com', password: 'Test1234', termsOfService: true })
+            .expect(201);
+    });
```
Comment (EricBanker12) [inline src/routes/auth/auth.routes.spec.ts]: expected 201 "Created", got 500 "Internal Server Error"

**Review by EricBanker12 — CHANGES_REQUESTED** [review]
There are a few problems with .env variables not having defaults for development. We'll also need Seeding, but that can be added with a separate branch in the future.

File: src/middleware/validate/validate.middleware.spec.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            }),
+        };
+        const next = jest.fn();
+
+        await Validation()(req, res, next);
+        expect(res.json).toHaveBeenCalled();
+        expect(res.json).toHaveBeenCalledWith({
+            errors: ['password must be longer than or equal to 8 characters'],
+        });
+    });
```
Comment (EricBanker12) [inline src/middleware/validate/validate.middleware.spec.ts]:
```
Validation() › should return a error if password is not at least 8 characters

- Expected
    + Received

      Object {
        "errors": Array [
    +     "email must be an email",
          "password must be longer than or equal to 8 characters",
        ],
      },
```
Other tests are failing too. I believe this is caused by `@IsEmail()` in `register.dto` always failing.

File: src/models/dto/register.dto.ts [hunk 6]
```diff
@@ -0,0 +1,16 @@
+import { IsEmail, IsString, Length, IsBoolean } from 'class-validator';
+import { LoginDTO } from './login.dto';
+
+class RegisterDTO extends LoginDTO {
+    @IsEmail()
```
Comment (EricBanker12) [inline src/models/dto/register.dto.ts:5]: Always fails in testing

**Review by EricBanker12 — CHANGES_REQUESTED** [review]
Email validation always fails in testing, even when it shouldn't.

File: src/middleware/validate/validate.middleware.spec.ts (hunk 5 above)
Comment (EricBanker12) [reply src/middleware/validate/validate.middleware.spec.ts]: either `body: { username: 'Test@mail.com',` should be `body: { email: 'Test@mail.com',` or dto need to be changed

**Review by EricBanker12 — APPROVED** [review]
Testing passes and manual tests work as expected.

**Pull request merged**